ambry-cloud/src/main/java/com.github.ambry.cloud/CloudBlobStore.java

```diff
@@ -368,20 +446,52 @@ private void checkStarted() throws StoreException {
   private void checkDuplicates(MessageWriteSet writeSet) {
     List<MessageInfo> infos = writeSet.getMessageSetInfo();
     if (infos.size() > 1) {
-      Set<StoreKey> seenKeys = new HashSet<>();
-      for (MessageInfo info : infos) {
-        if (!seenKeys.add(info.getStoreKey())) {
-          throw new IllegalArgumentException("WriteSet contains duplicates. Duplicate detected: " + info.getStoreKey());
-        }
+      List<StoreKey> keys = infos.stream().map(info -> info.getStoreKey()).collect(Collectors.toList());
+      Set<StoreKey> duplicates = getDuplicates(keys);
+      if (duplicates.size() > 0) {
+        throw new IllegalArgumentException("WriteSet contains duplicates. Duplicates detected: " + duplicates);
+      }
+    }
+  }
+
+  /**
+   * Detects duplicates in {@code keys}
+   * @param keys list of {@link StoreKey} to detect duplicates in
+   * @throws IllegalArgumentException if a duplicate is detected
+   */
+  private void checkDuplicates(List<? extends StoreKey> keys) {
+    if (keys.size() > 1) {
+      Set<StoreKey> duplicates = getDuplicates(keys);
+      if (duplicates.size() > 0) {
+        throw new IllegalArgumentException("StoreKey list contains duplicates. Duplicates detected: " + duplicates);
       }
     }
   }
 
+  /**
+   * Get duplicates in a list of {@code StoreKey}
+   * @param keys keys list of {@link StoreKey} to find duplicates in
+   * @return {@code Set} of duplicate {@code StoreKey}
+   */
+  private Set<StoreKey> getDuplicates(List<? extends StoreKey> keys) {
```

Comment:
Do we need this method?  I think we can move the logic to checkDuplicates(List<StoreKey>) and call that from checkDuplicates(MessageWriteSet)

Reply:
fixed.